### DISCUSSION TITLE
withCount @beforeFetch

### DISCUSSION THREAD

**Discussion by DealerUp** [adonisjs/core#2108] [Ideas]
Hi.

I implemented softDeletes using the example shown here  (@Obapelumi):
https://github.com/adonisjs/core/discussions/1791#discussioncomment-139879

Everything works perfectly fine! But the withCount method is not applying softDeletes hooks.
SoftDeletes.ts [Trait]
```typescript
export const softDeleteQuery = (query: ModelQueryBuilderContract<typeof BaseModel>) => {
  const table = query.model.table
  query.whereNull(`${table}.deleted_at`)
}
```

Order.ts [Model]
```typescript
@hasMany(() => OrderItem)
public items: HasMany<typeof OrderItem>
```

OrderItem.ts [Model]
```typescript
@beforeFetch()
public static softDeletesFetch = softDeleteQuery
```

OrdersController.ts [Controller]
```typescript
const order = await Order.query()
    .withCount('items', (query) => {
        query.sum('total_price').as('itemsAmount')
    })
```

I reviewed the code of the withCount method (@adonisjs/lucid/build/src/Orm/QueryBuilder/index.js) and after a lot of fighting, I applied a "solution" made in Brazil ;-).

The solution (before run select):
``` typescript
/**
 * Apply model hooks
 */
const applyHooks = async () => {
    await subQuery.model.$hooks.exec('before', 'fetch', subQuery);
}

applyHooks();
```
![image](https://user-images.githubusercontent.com/39589009/104028530-9704e200-51a7-11eb-963e-60dd33e4b852.png)

Is it possible and feasible to add this solution?
Is my analysis correct that I should apply hooks in this scenario?

Tks 4 all

**Comment by DealerUp** [comment]
@thetutlage any idea?

**Comment by thetutlage** [comment]
Hooks are not the way to go here. We will have proper support for global scopes, which serve the use case of softDeletes.

**Reply by DealerUp** [reply]
When will there be adequate support? Do you have any predictions?

**Reply by Obapelumi** [reply]
I've switched to using normal query scopes for my soft delete use cases because it is more flexible.
Depending on your use case you may not need global scopes. For example, if you decide to implement a restore delete feature then you could do:

```
export const softDeleteQuery = (
  query: ModelQueryBuilderContract<LucidModel>,
  isDeleted = true
) => {
  const column = `${query.model.table}.deleted_at`
  if (isDeleted) {
    query.whereNotNull(column)
  } else {
    query.whereNull(column)
  }
}
```
In your model

```
  public static deleted = scope(softDeleteQuery)
```

This way you can apply the scope to fetch either deleted or not deleted rows like so:

```
// not deleted
query.apply(scopes => scopes.deleted(false))

// deleted
query.apply(scopes => scopes.deleted())
```
